CONTRIBUTING.md

```diff
@@ -66,6 +66,111 @@ For first-time contributers, we suggest to look for issues categorized as [trivi
 
 If you want a hassle-free contribution look for issues categorized as [preapproved](https://issues.dlang.org/buglist.cgi?component=dmd&keywords=preapproved&product=D).
 
+## DMD Style
+
+Here is a short list of stylistic issues the core team will expect in
+pull requests. Much of the source code does not follow these, but
+we expect new code to, and PRs retrofitting the existing code to
+follow it is welcome.
+
+1. Use attributes `const`/`nothrow`/`pure`/`scope`/`@safe`/`private`/etc.
+Successfully using `pure` functions is regarded with particular favor.
+
+2. Use correct Ddoc function comment blocks. Do not use Ddoc comments for
+overrides unless the overriding function does something different (as far as
+the caller is concerned) than the overridden function. Ddoc comment blocks
+are often overkill for nested functions and function literals; use ordinary
+comments for those.
+
+3. Do not use `strlen`/`strcmp` and their ilk. Use D arrays instead.
+
+4. Use `ref` instead of raw pointers.
```

Comment:
`ref` should pretty much always be used where it can be used instead of a pointer.

Reply:
Unless it's `const ref`, I avoid `ref` because it means the caller can't tell that its value could be modified.

```D
int x = 42;
foo(x);
// what is x?
```

Since I avoid non-const `ref`, I know that `x` is still 42.  I don't have to worry about the immediate value of my parameters changing when I pass them to functions.

C# also supports ref, but requires that the caller specify it as well:
```csharp
int x = 42;
foo(x);
// x is still 42
bar(ref x);
// now we know x may have been modified
```

This is a big disadvantage in my book as the immediate value of every single value I pass to a function can now change from underneath me.  It adds a very large mental burden on the developer.